tests/test_util.py

```diff
@@ -0,0 +1,134 @@
+    @mock.patch("pythonforandroid.util.mkdtemp")
+    def test_temp_directory(self, mock_mkdtemp, mock_shutil_rmtree):
+        mock_mkdtemp.return_value = "/temp/any_directory"
+        with util.temp_directory():
+            mock_mkdtemp.assert_called_once()
+        mock_shutil_rmtree.assert_called_once_with("/temp/any_directory")
+
+    @mock.patch("pythonforandroid.util.chdir")
+    def test_current_directory(self, moch_chdir):
+        chdir_dir = "/temp/any_directory"
+        # test chdir to existing directory
+        with util.current_directory(chdir_dir):
+            moch_chdir.assert_called_once_with("/temp/any_directory")
+        moch_chdir.assert_has_calls(
+            [
+                mock.call("/temp/any_directory"),
+                mock.call(os.getcwd()),
+            ]
+        )
+
+    def test_current_directory_exception(self):
+        # test chdir to non-existing directory, should raise error
+        # for py3 the exception is FileNotFoundError and IOError for py2, to
+        # avoid introduce conditions, we test with a more generic exception
+        with self.assertRaises(OSError):
+            with util.current_directory("/fake/directory"):
+                # the line below will never be executed
+                print("")
+
+    @mock.patch("pythonforandroid.util.sh.which")
+    def test_get_virtualenv_executable(self, mock_sh_which):
+        # test that all calls to `sh.which` are performed, so we expect the
+        # first two `sh.which` calls should be None and the last one should
+        # return the expected virtualenv (the python3 one)
+        expected_venv = os.path.join(
+            os.path.expanduser("~"), ".local/bin/virtualenv"
+        )
+        mock_sh_which.side_effect = [None, None, expected_venv]
+        self.assertEqual(util.get_virtualenv_executable(), expected_venv)
+        mock_sh_which.assert_has_calls(
+            [
+                mock.call("virtualenv2"),
+                mock.call("virtualenv-2.7"),
+                mock.call("virtualenv"),
+            ]
+        )
+        self.assertEqual(mock_sh_which.call_count, 3)
+        mock_sh_which.reset_mock()
+
+        # Now test that we don't have virtualenv installed, so all calls to
+        # `sh.which` should return None
+        mock_sh_which.side_effect = [None, None, None]
+        self.assertIsNone(util.get_virtualenv_executable())
+        self.assertEqual(mock_sh_which.call_count, 3)
+        mock_sh_which.assert_has_calls(
+            [
+                mock.call("virtualenv2"),
+                mock.call("virtualenv-2.7"),
+                mock.call("virtualenv"),
+            ]
+        )
+
+    def test_walk_valid_filens_sample(self):
+        file_ens = util.walk_valid_filens(
+            "/home/opacam/Devel/python-for-android/tests/",
+            ["__pycache__"],
+            ["*.pyc"],
+        )
+        for i in os.walk("/home/opacam/Devel/python-for-android/tests/"):
+            print(i)
```

Comment:
Not sure if I'm missing something here, what is the value in printing this? There are no other checks?

Reply:
Yes this is definitely some debugging left over. Probably the while test should go

Reply:
Thanks, already removed